File: python/tests/DWX_Performance_Test.py

import sys
import json
from time import sleep
from threading import Thread
from os.path import join, exists
from traceback import print_exc
from random import random
from datetime import datetime, timedelta

sys.path.append('../')
from api.DWX_Client import DWX_Client


"""

Performance Test

This test will measure how long it takes to open, modify and close pending orders. 
It will open 100 pending orders and calculate the average durations. 

Please don't run this on your live account. 

The MT4/5 server must be initialized with MaximumOrders>=100. 


"""

class TickProcessor():

    def __init__(self, MT4_directory_path, 
                 sleep_delay=0.005,             # 5 ms for time.sleep()
                 max_retry_command_seconds=10,  # retry to send the commend for 10 seconds if not successful. 
                 verbose=True
                 ):

        self.last_open_time = datetime.utcnow()
        self.last_modification_time = datetime.utcnow()

        self.dwx = DWX_Client(self, MT4_directory_path, sleep_delay, 
                              max_retry_command_seconds, verbose=verbose)
        sleep(1)

        self.dwx.start()
        
        self.n = 100
        symbol = 'EURUSD'
        entry_price = 1.17

        self.test_started = False
        self.n_modified = 0
        self.dwx.close_all_orders()
        while(len(self.dwx.open_orders) != 0):
            # print(self.dwx.open_orders)
            sleep(1)

        self.test_started = True
        self.before_open = datetime.utcnow()
        self.open_duration = -100
        for i in range(self.n):
            self.dwx.open_order(symbol=symbol, order_type='buylimit', lots=0.01, price=entry_price)
        
        
        while(len(self.dwx.open_orders) < self.n):
            sleep(1)
        
        self.modify_duration = -100
        self.before_modification = datetime.utcnow()
        for ticket in self.dwx.open_orders.keys():
            self.dwx.modify_order(ticket, stop_loss=entry_price-0.01)
        
        sleep(1)

        self.before_close = datetime.utcnow()
        for ticket in self.dwx.open_orders.keys():
            self.dwx.close_order(ticket)
        

    def on_tick(self, symbol, bid, ask):

        now = datetime.utcnow()

        print('on_tick:', now, symbol, bid, ask)


    def on_bar_data(self, symbol, time_frame, time, open_price, high, low, close_price, tick_volume):
        
        print('on_bar_data:', symbol, time_frame, datetime.utcnow(), time, open_price, high, low, close_price)


    def on_historic_data(self, symbol, time_frame, data):
        
        # you can also access the historic data via self.dwx.historic_data. 
        print('historic_data:', symbol, time_frame, f'{len(data)} bars')


    def on_message(self, message):

        if message['type'] == 'ERROR':
            print(message['type'], '|', message['error_type'], '|', message['description'])
        elif message['type'] == 'INFO':
            print(message['type'], '|', message['message'])
            if 'modified' in message['message']:
                self.n_modified += 1
                if self.n_modified == self.n:
                    self.modify_duration = (datetime.utcnow() - self.before_modification).total_seconds()


    # triggers when an order is added or removed, not when only modified. 
    def on_order_event(self):
        
        print(f'on_order_event. open_orders: {len(self.dwx.open_orders)} open orders')

        if not self.test_started:
            return
        
        if len(self.dwx.open_orders) == self.n:
            self.open_duration = (datetime.utcnow() - self.before_open).total_seconds()
            
        elif len(self.dwx.open_orders) == 0:
            close_duration = (datetime.utcnow() - self.before_close).total_seconds()
            print(f'\nopen_duration: {1000*self.open_duration/self.n:.1f} milliseconds per order')
            print(f'modify_duration: {1000*self.modify_duration/self.n:.1f} milliseconds per order')
            print(f'close_duration: {1000*close_duration/self.n:.1f} milliseconds per order')


mt4_files_path = 'C:/Users/Administrator/AppData/Roaming/MetaQuotes/Terminal/3B534B10135CFEDF8CD1AAB8BD994B13/MQL4/Files/'
processor = TickProcessor(mt4_files_path)

while processor.dwx.ACTIVE:
    sleep(1)

File: java/tests/DWXPerformanceTest.java
import org.json.JSONArray;
import org.json.JSONObject;
import java.util.ArrayList;
import java.util.Arrays;
import org.junit.Assert;
import org.junit.Test;
import org.junit.Before;

import api.Client;
import api.EventHandler;

import static api.Helpers.*;



/*

Performance Test

This test will measure how long it takes to open, modify and close pending orders. 
It will open 100 pending orders and calculate the average durations. 

Please don't run this on your live account. 

The MT4/5 server must be initialized with MaximumOrders>=100. 


compile and run:

javac -cp ".;../;../libs/*" DWXPerformanceTest.java
java -cp ".;../;../libs/*" DWXPerformanceTest


JSON jar from here:
https://github.com/junit-team/junit4/wiki/Download-and-Install

*/

public class DWXPerformanceTest {
    
	final static String MetaTraderDirPath = "C:/Users/asd/AppData/Roaming/MetaQuotes/Terminal/3B534B10135CFEDF8CD1AAB8BD994B13/MQL4/Files/";
	// final static String MetaTraderDirPath = "D:/MetaTrader5_portable/MQL5/Files/";
	final static int sleepDelay = 5;  // 5 milliseconds
	final static int maxRetryCommandSeconds = 10;
	final static boolean loadOrdersFromFile = true;
	final static boolean verbose = true;
	
	
    public static void main(String args[]) throws Exception {
        
        MyEventHandler eventHandler = new MyEventHandler();
        
        Client dwx = new Client(eventHandler, MetaTraderDirPath, sleepDelay, 
                                maxRetryCommandSeconds, loadOrdersFromFile, verbose);
        
    }
}

class MyEventHandler implements EventHandler {
	
	int n = 100;
    String symbol = "EURUSD";
    double entryPrice = 1.17;
	
	long beforeOpen = System.currentTimeMillis();
    long openDuration = -100;
	boolean testStarted = false;
	int nModified = 0;
	long modifyDuration = -100;
    long beforeModification = System.currentTimeMillis();
	long beforeClose = System.currentTimeMillis();
	
    
    public void start(Client dwx) {
        
		print("\nAccount info:\n" + dwx.accountInfo + "\n");
		
        dwx.closeAllOrders();
        while(dwx.openOrders.length() != 0)
            sleep(1000);

        testStarted = true;
        
        for (int i=0; i<n; i++)
            dwx.openOrder(symbol, "buylimit", 0.01, entryPrice, 0, 0, 0, "", 0);
        
        
        while(dwx.openOrders.length() < n)
            sleep(1000);
        
        beforeModification = System.currentTimeMillis();
		for (String ticket : dwx.openOrders.keySet())
            dwx.modifyOrder(Integer.valueOf(ticket), 0.01, 0, entryPrice-0.01, 0, 0);
        
        sleep(1000);

        beforeClose = System.currentTimeMillis();
        for (String ticket : dwx.openOrders.keySet())
            dwx.closeOrder(Integer.valueOf(ticket));
    }
    
	
    public synchronized void onTick(Client dwx, String symbol, double bid, double ask) {
        
		print("onTick: " + symbol + " | bid: " + bid + " | ask: " + ask);
    }
	
    
    public synchronized void onBarData(Client dwx, String symbol, String timeFrame, String time, double open, double high, double low, double close, int tickVolume) {
        
		print("onBarData: " + symbol + ", " + timeFrame + ", " + time + ", " + open + ", " + high + ", " + low + ", " + close + ", " + tickVolume);
    }
	
	
	public synchronized void onHistoricData(Client dwx, String symbol, String timeFrame, JSONObject data) {
        
		// you can also access historic data via: dwx.historicData
		print("onHistoricData: " + symbol + ", " + timeFrame + ", " + data);
    }
	
    
    public synchronized void onMessage(Client dwx, JSONObject message) {
		
        if (message.get("type").equals("ERROR")) {
			print(message.get("type") + " | " + message.get("error_type") + " | " + message.get("description"));
		} else if (message.get("type").equals("INFO")) {
			print(message.get("type") + " | " + message.get("message"));
			
			if (((String)message.get("message")).contains("modified")) {
                nModified++;
                if (nModified == n)
                    modifyDuration = System.currentTimeMillis() - beforeModification;
			}
		}
    }
	
    
    // triggers when an order is added or removed, not when only modified. 
    public synchronized void onOrderEvent(Client dwx) {
		
        print("onOrderEvent:");
        
        // dwx.openOrders is a JSONObject, which can be accessed like this:
        for (String ticket : dwx.openOrders.keySet()) 
            print(ticket + ": " + dwx.openOrders.get(ticket));
		
		
		if (!testStarted)
            return;
        
        if (dwx.openOrders.length() == n) {
            openDuration = System.currentTimeMillis() - beforeOpen;
        } else if (dwx.openOrders.length() == 0) {
            long closeDuration = System.currentTimeMillis() - beforeClose;
            print("\nopenDuration: " + openDuration/n + " milliseconds per order");
            print("modifyDuration: " + modifyDuration/n + " milliseconds per order");
            print("closeDuration: " + closeDuration/n + " milliseconds per order");
		}
    }
}

